### PR TITLE
Agent history compaction: one processEvent arm on the token-usage machinery

### PR DESCRIPTION
**Pull request by jonastemplestein**
Stacked on #1809 (the token-counting + history-rewriting machinery). This PR is the payoff, and deliberately small: **compaction is one processEvent arm and a summarizer prompt.**

The machinery PR gives every agent stream two facts: `agent/token-usage-reported` (how full did that turn run, against which window) and `agent/history-reset` (replace the model-visible conversation wholesale). Compaction is then just a reaction connecting them:

- the arm fires on every usage report; past **half the model window** (`AGENT_COMPACTION_TRIGGER_FRACTION`) it starts a droppable background attempt,
- the attempt summarizes the whole conversation with the agent's **own model** (`AGENT_COMPACTION_PROMPT` optimizes for retrieval keys — names, paths, ids, open tasks — over narrative),
- and appends the `history-reset` itself: summary as a single user-role item, turns that landed mid-summary carried forward verbatim, `reason: "compaction@<offset>: ~N tokens > M"`.

Correctness posture mirrors the LLM turn's (droppable attempt + durable evidence): the reset's idempotency key derives from the triggering report's offset, so a redelivered trigger finds the existing reset and skips before the AI call; a newer reset anywhere after the trigger makes it stale; a rewritten history prefix (concurrent reset won) drops the summary; concurrent triggers serialize on a per-incarnation chain. A failed summary changes nothing durable — the next over-threshold report is the retry. We are deliberately NOT solving every edge case here; the point is the shape.

Tests: the full loop (over-threshold report → summary on the agent's model → reset with the summary → a fresh incarnation redelivering the whole journal does NOT summarize again) and the under-threshold no-op.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Compaction rewrites model-visible history and blocks delivery during summarization, so correctness and timing around concurrent inputs and LLM turns matter; scope is confined to the agent processor with idempotent guards and swallowed summary errors.
> 
> **Overview**
> Adds **automatic conversation compaction** when a completed turn’s `token-usage-reported` payload shows input+output tokens at or above **50%** of `maxContextTokens` (`AGENT_COMPACTION_TRIGGER_FRACTION`).
> 
> On trigger, the processor runs a **stop-the-world** path via `blockProcessorWhile`: it calls Workers AI with the agent’s configured model and `AGENT_COMPACTION_PROMPT`, then appends `agent/history-reset` with a summary user message, preserves the system prompt, and **carries forward** any history that arrived during the summary. Idempotency is keyed off the usage report offset; a reset after the measured LLM request skips re-compaction on replay. Failures are best-effort (no journal change; next over-threshold report retries). Contract/docs now describe the processor-owned compaction lane and list `history-reset` among emitted events.
> 
> Tests cover the full loop (over-threshold → one summary call → reset → no second compaction on journal replay) and the under-threshold no-op.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit e5b604d29a2d701cde71e258a41069f80cf7c5e7. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

| Group | Lines | Significant |
| --- | ---: | ---: |
| Product | +141 -3 | +94 -3 🟩🟩🟩🟩🟩 |
| Tests | +118 -0 | +103 -0 🟩🟩🟩🟩🟩 |
| **Total** | **+259 -3** | **+197 -3** 🟩🟩🟩🟩🟩 |

<sub>Lines counts every changed line; Significant ignores blank lines and JS comments. Between a2f5c6f and e5b604d, bucketed first-match-wins into the groups defined in `scripts/ci/loc-report.ts`.</sub>

## Environment Config Lease

<details>
<summary>No active environment config lease.</summary>

| app | status | commit | preview | size (gzip) | deploy duration | test duration | retries | cleanup duration | workflow run | updated | summary |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
| Auth | released | `e5b604d` | [https://auth.iterate-preview-8.com](https://auth.iterate-preview-8.com) | 819.1 KiB | 19.1s | 544ms |  | 3.9s | [Workflow run](https://github.com/iterate/iterate/actions/runs/342864862077634) | 2026-07-10T11:29:11.067Z | Preview app released. |
| OS | released | `e5b604d` | [https://os.iterate-preview-8.com](https://os.iterate-preview-8.com) | 4.16 MiB (+1.6 KiB vs main) | 93.1s | 122.3s | 3 retried: DESIRED: a live-capability fetch handler serves WebSockets at an app host (vitest x1) · cross-post chains accumulate provenance hops and removeCrossPost stops forwarding (vitest x1) · can enter the dashboard with a forged session (specs x1) | 15.4s | [Workflow run](https://github.com/iterate/iterate/actions/runs/342864862077634) | 2026-07-10T11:29:22.497Z | Preview app released. |
| Semaphore | released | `2791eeb` | [https://semaphore.iterate-preview-8.com](https://semaphore.iterate-preview-8.com) | 440.8 KiB | 15.3s | 6.2s |  | 797ms | [Workflow run](https://github.com/iterate/iterate/actions/runs/502381042025282) | 2026-07-10T11:29:07.943Z | Preview app released. |
| Streams Example App | released | `5bf9500` | [https://streams.iterate-preview-8.com](https://streams.iterate-preview-8.com) | 1.33 MiB | 13.6s | 16.1s |  | 1.0s | [Workflow run](https://github.com/iterate/iterate/actions/runs/95731025175971) | 2026-07-10T11:29:08.183Z | Preview app released. |

</details>

### CODE REVIEW

File: apps/os/src/domains/agents/agent-processor-implementation.ts [hunk 1]
```diff
@@ -291,11 +292,121 @@ export class AgentProcessor extends StreamProcessor<AgentProcessorContract, Agen
...
+          history: [
+            {
+              role: "user" as const,
+              content: `[Earlier conversation history was compacted. Summary:]\n\n${summary.text}`,
+            },
+            ...carriedForward,
+          ],
+          reason: `compaction@${triggerOffset}: ~${contextTokens} tokens > ${thresholdTokens}`,
+        },
+      });
```
[the remark this answers is not in the document]
Comment (jonastemplestein) [reply apps/os/src/domains/agents/agent-processor-implementation.ts:407]: Fixed by reworking compaction to stop-the-world (2e3d490b6): the trigger now runs under `blockProcessorWhile`, so the processor delivers nothing — and starts no turns — until the reset lands; the agent's own outputs can no longer commit into the gap. External messages that land mid-summary are carried forward behind the summary in the reset payload (queued input, processed after compaction). The instant between the carried-forward read and the append is irreducible without a compare-and-append primitive and is accepted for this best-effort lane: a message in that sliver stays in the journal and the UI, and only drops out of the model-visible view.
